packages/gatsby/src/bootstrap/cache/remove.js

```diff
@@ -0,0 +1,70 @@
+const path = require(`path`)
+const fs = require(`fs-extra`)
+const _ = require(`lodash`)
+
+const getPluginHash = require(`./plugin-hash`)
+
+const CACHE_FOLDER = `caches`
+const DONT_DELETE = [`redux.state`, `loki`]
+/*
+ * This function will first calculate a diff of plugin/file changes (e.g. if gatsby-node.js changes)
+ * From this, it will then selectively invalidate files in .cache/caches
+ * the location where Gatsby stores long-lived cache files that can be safely kept around
+ */
+module.exports = async function safeRemoveCache({
+  directory,
+  plugins,
+  existing,
+  report,
+}) {
+  const cacheDirectory = path.join(directory, `.cache`)
+
+  const { changes, hash } = await getPluginHash({
+    plugins,
+    existing,
+  })
+
+  const equalHashes = _.isEqual(existing, hash)
+  const hasExistingCache = Object.keys(existing).length > 0
+
+  await fs.ensureDir(cacheDirectory)
+
+  if (hasExistingCache && !equalHashes) {
+    report.info(report.stripIndent`
+      The following plugins were changed, added or removed since the last time you ran Gatsby:
+${changes.map(change => `        - ${change}`).join(`\n`)}
+      As a precaution, we're invalidating parts of your cache to ensure your data is sparkly fresh.
+    `)
+  }
+
+  if (!equalHashes) {
+    try {
+      const changesLookup = changes.reduce((merged, folderName) => {
+        merged[folderName] = true
+        return merged
+      }, {})
+      const filesToRemove = await fs.readdir(cacheDirectory).then(allFiles =>
```

Comment:
note to self - this won't work for plugins that have `/` in name - for example:
- `@scope/name`
- `nested/local/plugin`

(it assumes flat directory structure)

Reply:
ops, it should be comment to line 50, but same thing applies